programs/leap-util/actions/blocklog.cpp

```diff
@@ -47,117 +47,149 @@ struct report_time {
-               return -1;
-         }
-         return 0;
-      }
-      if(opt->extract_blocks) {
-         if(opt->first_block == 0 && opt->last_block == std::numeric_limits<uint32_t>::max()) {
-            std::cerr << "extract-blocklog does nothing unless first and/or last block are specified.";
-            return -1;
-         }
-         if(!extract_block_range(opt->blocks_dir, opt->output_dir, opt->first_block, opt->last_block))
-            return -1;
-         return 0;
-      }
-      if(opt->vacuum) {
-         initialize();
-         return do_vacuum();
-      }
-      if(opt->genesis) {
-         initialize();
-         return do_genesis();
-      }
-      if(opt->make_index) {
-         const bfs::path blocks_dir = opt->blocks_dir;
-         bfs::path out_file = blocks_dir / "blocks.index";
-         const bfs::path block_file = blocks_dir / "blocks.log";
-
-         if(!opt->output_file.empty())
-            out_file = opt->output_file;
-
-         report_time rt("making index");
-         const auto log_level = fc::logger::get(DEFAULT_LOGGER).get_log_level();
-         fc::logger::get(DEFAULT_LOGGER).set_log_level(fc::log_level::debug);
-         block_log::construct_index(block_file.generic_string(), out_file.generic_string());
-         fc::logger::get(DEFAULT_LOGGER).set_log_level(log_level);
-         rt.report();
-         return 0;
-      }
-      //else print blocks.log as JSON
-      initialize();
-      read_log();
+      throw;
    } catch(const fc::exception& e) {
       elog("${e}", ("e", e.to_detail_string()));
-      return -1;
    } catch(const boost::exception& e) {
       elog("${e}", ("e", boost::diagnostic_information(e)));
-      return -1;
+   } catch(const CLI::RuntimeError& e) {
+      // avoid reporting it twice, RuntimeError is only for cli11
    } catch(const std::exception& e) {
       elog("${e}", ("e", e.what()));
-      return -1;
    } catch(...) {
       elog("unknown exception");
-      return -1;
    }
+}
+
+void blocklog_actions::initialize() {
+   try {
+      bfs::path bld = opt->blocks_dir;
+      if(bld.is_relative())
+         opt->blocks_dir = (bfs::current_path() / bld).string();
+      else
+         opt->blocks_dir = bld.string();
+
+      if(!opt->output_file.empty()) {
+         bld = opt->output_file;
+         if(bld.is_relative())
+            opt->output_file = (bfs::current_path() / bld).string();
+         else
+            opt->output_file = bld.string();
+      }
+
+      //if the log is pruned, keep it that way by passing in a config with a large block pruning value. There is otherwise no
+      // way to tell block_log "keep the current non/pruneness of the log"
+      if(block_log::is_pruned_log(opt->blocks_dir)) {
+         opt->blog_keep_prune_conf.emplace();
+         opt->blog_keep_prune_conf->prune_blocks = UINT32_MAX;
+      }
+   }
+   FC_LOG_AND_RETHROW()
+}
+
+int blocklog_actions::make_index() {
+   initialize();
```

Comment:
Every subcommand calls `initialize`. Can it be called once in some common place?